src/tqec/compile/substitute.py

```diff
@@ -0,0 +1,218 @@
+    # Substitute the first layer of the top block with the bulk layer
+    top_block = block2.with_updated_layer(
+        plaquettes_to_update=deepcopy(block2.layers[1].collection),
+        layer_to_update=0,
+    )
+    # If the pipe has a hadamard, apply the hadamard transformation
+    # The hadamard transformation is applied to the end of the bottom block
+    if key.pipe_type.has_hadamard:
+        for plaquette in bottom_block.layers[-1].collection.values():
+            _inplace_append_timelike_hadamard_transition(plaquette)
+    return bottom_block, top_block
+
+
+def _inplace_append_timelike_hadamard_transition(
+    plaquette: Plaquette,
+) -> None:
+    """Append a moment of transversal hadamard gates to the data qubits in the
+    plaquette."""
+    max_schedule_index = max(plaquette.circuit.schedule)
+    plaquette.circuit.add_to_schedule_index(
+        schedule=max_schedule_index + 1,
+        moment=cirq.Moment(
+            cirq.H(dq.to_grid_qubit()).with_tags(Plaquette._MERGEABLE_TAG)
+            for dq in plaquette.qubits.data_qubits
+        ),
+    )
+
+
+def _substitute_in_space_with_usual_cubes(
+    key: SubstitutionKey,
+    block1: CompiledBlock,
+    block2: CompiledBlock,
+) -> tuple[CompiledBlock, CompiledBlock]:
+    pipe_type = key.pipe_type
+    assert (
+        not key.spec1.is_spatial_junction and not key.spec2.is_spatial_junction
+    ), "Both cubes must not be spatial junction."
+    assert pipe_type.direction != Direction3D.Z, "Pipe direction must be X or Y."
+    # TODO: If the pipe has a hadamard, apply the spatial hadamard transformation
+    if pipe_type.has_hadamard:
+        raise TQECException("Hadamard on spatial pipes is not supported yet.")
+    substitute_side1 = (
+        PlaquetteSide.RIGHT
+        if pipe_type.direction == Direction3D.X
+        else PlaquetteSide.DOWN
+    )
+    substitute_side2 = substitute_side1.opposite()
+    # Get reset/mesurement basis
+    basis = pipe_type.value[2].upper()
+    block1 = _substitute_on_the_border(block1, substitute_side1, basis)
+    block2 = _substitute_on_the_border(block2, substitute_side2, basis)
+    return block1, block2
+
+
+def _substitute_on_the_border(
+    block: CompiledBlock,
+    plaquette_side: PlaquetteSide,
+    basis: str,
+) -> CompiledBlock:
+    if plaquette_side == PlaquetteSide.LEFT:
+        substitution = {1: 6, 8: 9, 7: 10}
+    elif plaquette_side == PlaquetteSide.UP:
+        substitution = {2: 12, 5: 10, 6: 9}
+    elif plaquette_side == PlaquetteSide.DOWN:
+        substitution = {3: 7, 14: 10, 13: 9}
+    else:  # plaquette_side == PlaquetteSide.RIGHT:
+        substitution = {4: 13, 11: 9, 12: 10}
+    # 1. Fill in/Substitue the plaquettes on the border
+    for i in range(block.num_layers):
+        block = block.with_updated_layer(
+            plaquettes_to_update={
+                dst: deepcopy(block.layers[i][src]) for dst, src in substitution.items()
+            },
+            layer_to_update=i,
+        )
```

Comment:
Seems like you pick plaquettes from the block to replace other plaquettes from the block. I am still convinced that we need to take the plaquettes from independent pipe instances, that are defined just like blocks (except that they only have 2 layers).

See [Austin's answer](https://github.com/QCHackers/tqec/pull/298#issuecomment-2327470006) to [my remark](https://github.com/QCHackers/tqec/pull/298#issuecomment-2327470006).

Reply:
I don't think it will be a problem since we can define the plaquettes for substitution arbitrarily. For example, to implement a spatial hadamard pipe, we can define new types of plaquette(tweaked half x half z) easily, or even only use some substitution rules to update the circuit in the plaquette(like replace the two CX on a side with two CZ). 

Instead, if we define the layers within a pipe, the same plaquettes in a specific type of pipe may not always stay valid for substitution. For example, a spatial pipe connects to a junction may apply different substitution rules compared with connecting to a usual cube(I guess)...

Reply:
After watching the recording, I agree that it's better to implement the pipes as templates and plaquette layers. The main drawback of the simple plaquettes substitution approach is that we can not change the underlying template.

Reply:
Thinking about it, we should be able to use the same "spec" mechanism you implemented for pipes, I think it would even generalise to special pipes such as the ones used in spatial junctions to keep the logical qubit shape (did we call that parity at one point?).

Reply:
Do you have a situation in mind where we might have to change the underlying template?

Reply:
> Thinking about it, we should be able to use the same "spec" mechanism you implemented for pipes, I think it would even generalise to special pipes such as the ones used in spatial junctions to keep the logical qubit shape.

I agree that the "spec" mechanism can be used for pipes and this is the idea behind `SubstitutionKey`.

> did we call that parity at one point?

I suppose you mean the odd/even boundary plaquette placement. The word "parity" seems ambiguous to me since it can mean a lot in qec like parity of checks.

> Do you have a situation in mind where we might have to change the underlying template?

I thought for a while and nothing came to my mind... So maybe it's not a problem.